tools/fix_mailmap.py

```diff
@@ -0,0 +1,573 @@
+    """Print messages indicating changes were made that the user should confirm."""
+    global changes_made
+    changes_made = True
+    if quiet_mode:
+        return
+    if color_output:
+        text = "\n".join(f"\033[33m{line}\033[0m" for line in text.splitlines())
+    print(text, file=stderr)
+
+
+def print_green(text: str) -> None:
+    """Print messages indicating no changes needed."""
+    if quiet_mode:
+        return
+    if color_output:
+        text = "\n".join(f"\033[32m{line}\033[0m" for line in text.splitlines())
+    print(text, file=stderr)
+
+
+def get_mailmap_path() -> Path:
+    """Use git to to find the relative path to the .mailmap file."""
+    cmd = ["git", "rev-parse", "--show-cdup"]
+    res = subprocess.run(cmd, capture_output=True, check=True, text=True)
+    return Path(res.stdout.strip()) / ".mailmap"
+
+
+def get_raw_authors() -> dict[str, set[tuple]]:
+    """Get the list of authors WITHOUT .mailmap remapping.
+
+    Returns a dict {lowercase_email: set((name, email))}, which will be used to check if a
+    mailmap line is necessary (matches some commit, and actually makes a change)
+    """
+    cmd = [
+        "git",
+        "log",
+        "--pretty=tformat:%an <%ae>%n%(trailers:key=co-authored-by,valueonly=true,unfold=true)",
+    ]
+    res = subprocess.run(cmd, capture_output=True, check=True, text=True)
+
+    logline = re.compile(r"([^<]+)\s+<(.*)>$")
+
+    emails_map: dict[str, set[tuple]] = {}
+    for line in res.stdout.splitlines():
+        match = logline.fullmatch(line.strip())
+        if match:
+            name, email = match.group(1, 2)
+            emails_map.setdefault(email.lower(), set()).add((name, email))
+    return emails_map
+
+
+def get_shortlog_authors() -> tuple[list[tuple], list[str]]:
+    """Get the list of authors from git-shortlog (which uses .mailmap).
+
+    Returns (matches: (num_commits, name, email), non_matches)
+    """
+    cmd = ["git", "shortlog", "-se", "--group=author", "--group=trailer:co-authored-by"]
+    res = subprocess.run(cmd, capture_output=True, check=True, text=True)
+
+    matches = []
+    nonmatches = []
+    shortlogline = re.compile(r"^(\d+)\s+([^<]+)\s+<(.*)>$")
+    for line in res.stdout.splitlines():
+        match = shortlogline.fullmatch(line.strip())
+        if match:
+            matches.append((int(match.group(1)), *match.group(2, 3)))
+        else:
+            # These are usually faulty co-author lines missing <> around the email address eg
+            # Co-Authored-By: firstname lastname email@example.com
+            nonmatches.append(line)
+    return matches, nonmatches
+
+
+def split_mailmap_header_footer(mailmap_file: Path) -> tuple:
+    """Return the mailmap split into header comment lines and body lines."""
+    mailmap = mailmap_file.read_text(encoding="UTF8").splitlines()
+    for i, line in enumerate(mailmap):
+        if not line.startswith("#"):
+            header_end = i
+            break
```

Comment:
Did you mean to assign `header_end = 0` before the loop to avoid the possible `NameError`?  Otherwise there's no advantage over just calling the loop variable `header_end` directly.

Also, did you want to maybe include trailing blanks (without comments) in the header parsing?